1-Draft/RFCNNNN-Improve-Background-Operator.md

```diff
@@ -0,0 +1,216 @@
+```
+
+### Showing or hiding the output from one or more jobs
+
+This allows job output to be shown or hidden on demand for any jobs.
+
+```powershell
+# Show the output from job ids 1-5
+Show-JobOutput -Id 1..5
+# Hide the output from job name Job2
+Hide-JobOutput -Name Job2
+# Show the output for the last job (or jobs) we launched
+Show-JobOutput -Job $!
+```
+
+```output
+# As described in the comments above, with the output showing in an overlay
+```
+
+### Viewing the process ID for a job
+
+This shows how you can access the read-only process ID for a job.
+
+```powershell
+$!.ProcessId
+```
+
+```output
+# The process id for the background job(s)
+```
+
+## Specification
+
+1. For jobs invoked with `&!` or `Start-Job -ShowOutput`, or for jobs referenced
+in `Show-JobOutput`, hook up event handlers on the output stream collections for
+the job objects to capture the output from the jobs as it is added in real
+time, and show that output in the current terminal with the job name or ID in
+front of each output record. Mirror how `ProgressInformation` is rendered when
+it comes to showing job output this way.
+1. Add a `&!` operator that starts a job with the output display turned on.
+1. Add a boolean property to `BackgroundJob` objects called `ShowOutput` that
+identifies if the job is currently configured to show output or not.
+1. Add an integer property to `BackgroundJob` objects called `ProcessId` that
+identifies the ID of the process where the job is running.
+1. Add a `-ShowOutput` parameter to `Start-Job` that sets `ShowOutput` to true
+and hooks up event handlers as described in the first item in this list.
+1. Add a `$!` variable to store the most recently run job (or jobs if multiple
+jobs are launched at the same time, e.g. `cmd1 & cmd2 & cmd3 &`).
+
+## Alternate Proposals and Considerations
+
+### Add the process ID to the default output for a background job
+
+It may seem like a good idea to show the process ID for a background job in the
+default tabular output. While this would be convenient, we also have thread
+jobs that don't have a process ID, so it is probably better to leave it out for
+consistency across all jobs. It is always accessible via the `ProcessID` member
+of background jobs regardless.
+
+### Flip the silent/noisy approach with the operators
+
+This approach is proposed because we've already shipped the `&` background
+operator, with jobs running silent by default. In bash, if you run a command in
+the background using the `&` control operator, the output is displayed in the
+current terminal. This discrepancy may be confusing to users. If we feel that
+the PowerShell `&` background operator should be more consistent with the bash
+equivalent, we could make the `&` background operator show output by default
+and have the new `&!` background operator run silent. This wouldn't be a
+breaking change, but it would change the behavior for any scripts that are
```

Comment:
Absolutely accurate that formatting/output is not a breaking change, but the implications could be significant on machines with older console hosts running scripts that have a large number of jobs running "non-interactively" (ie in a real console host, but without anyone actually looking at the screen).

Reply:
Agreed. I think that alternative is a non-starter, and could just remove it. I only kept it because that was my original proposal, but then I flipped things around to persist current behavior and make this RFC additive rather than changing things.

Reply:
Removed the undesirable alternative.